proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+        From imports {imported from ‘Prelude’ at TTH.hs:3:8-11}
+      • In the Template Haskell typed quotation [|| show ||]
+    |
+  8 | foo = [|| show ||]
+    |
+
+
+The language of constraints is not yet expressive enough to communicate that we
+want the ``Show a`` evidence to be available at a later stage. Fixing this problem
+will require
+significant additional effort, and there are other known issues with TTH (see
+`Staging with Class: a Specification for Typed Template Haskell
+<https://dl.acm.org/doi/abs/10.1145/3498723>`_). We propose that an initial
+implementation of ``NoImplicitStagePersistence`` may support untyped TH but not
+TTH (i.e. the compiler may reject programs using TTH under
+``NoImplicitStagePersistence``).  In the long term, we believe that implementing
+Staging with Class is desirable and consistent with the direction of travel
+established by this proposal, but the full details of Staging with Class are out
+of scope.
+
+
+Implicit lifting and deriving ``Lift`` instances
+################################################
+.. _lift_instances:
+
+``Lift`` instances are used to provide serialisation-based cross-stage
+persistence.  For example, a typical ``Lift`` instance looks like::
+
+    data MInt = Some Int | None
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+The presence of this instance means the following declaration will be accepted::
+
+  foo :: MInt -> Q Exp
+  foo x = [| x |]  -- implicitly becomes [| $(lift x) |]
+
+Defining a ``Lift`` instance requires the datatype constructors to be available
+both at compile-time and runtime, so defining ``Lift`` within the same module as
+the datatype itself requires path-based cross-stage persistence.  Operationally,
+``None`` and ``Some`` are needed both at compile-time *and*  runtime since they
+are both matched on at compile time, and also persisted to be spliced in the
+future into a program that can make use of them at runtime. As a result, it
+isn't possible to define or derive a (non-orphan) ``Lift`` instance under
+``NoImplicitStagePersistence``.
+
+An orphan ``Lift`` instance can be defined thus::
+
+  module M where
+    data MInt = Some Int | None
+
+  module N where
+    import M
+    import quote M
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+This isn't technically problematic, rather it is just a result of what ``Lift``
+means. However, it means some users may need to modify their use of ``Lift``
+instances if they wish to benefit more from ``NoImplicitStagePersistence``.
+Users are free to use ``ImplicitStagePersistence`` in selected modules to allow
+defining ``Lift`` instances, but doing so means all the dependencies of the
+module will need to be available both at compile-time and runtime.
```

Comment:
I'm not following this argument. Suppose I enable `ExplicitLevelImports` and `ImplicitStagePersistence` as in the following example:

```hs
module A where { a = 1 :: Int }

module B where { b = 2 :: Int }

{-# LANGUAGE ExplicitLevelImports #-}
{-# LANGUAGE ImplicitStagePersistence #-}
module M where
import A
import quote B

data MInt = Some Int | None

instance Lift MInt where
        lift :: MInt -> Q Exp
        lift None     = [| None |]
        lift (Some x)  = if a == 1 then  [| Some (b + $(lift x)) |] else [| None |]

{-# LANGUAGE ExplicitLevelImports #-}
module C where
import M
import splice M
c :: MInt
c = $(lift (Some 42)) 
```

If `C` is required at stage R, then the `Lift MInt` instance is required at stage C.
Then I would expect that `A` is only needed at stage C as well, whereas `B` is only needed at stage R.

Both `A` and `B` are dependencies of module `M` which enables `ImplicitStagePersistence`, yet neither dependency is needed at both stages (although `M` *is* needed at both stages).

I don't like CSP, but it is perhaps less harmful than this proposal suggests.

Reply:
I could not follow. Could you perhaps give an example?

My main point is that CSP does not make it any harder to infer/check the levels at which an imported binding occurs, hence it should not affect the effectiveness of explicit level imports.

Reply:
@sgraf812 

Firstly:

If `C` is compiled at `R`, then by importing `M (Normal)` and `M (Splice)` you require `M @ R` and `M @ C`. 
Therefore you will need `A @ R` and `A @ C`, since `M` imports `A` without an offset.

Secondly, consider that you only `import splice M`, under the proposal because `A` using the ISP extension then we require `A @ C` and `A @ R`. `A @ R` is necessary as when you link you then you need the definition of `MInt` available to link against, as you have generated a program like `Some (b + 42)`.

Reply:
> Therefore you will need A @ R and A @ C

Indeed, thanks for the correction. Still, `B` is only needed at `R` despite it being a dependency of module `M` using the ISP extension, no? The proposal currently says that `B` will also be needed at `C`. I don't see why.

> A @ R is necessary as when you link you then you need the definition of MInt available to link against, as you have generated a program like Some (b + 42).

But `MInt` is defined in `M`, not `A`. The generated program does not mention `a` from module `A`, so I don't see why we need to link against `A`, and hence not why `A @ R` is needed in this second scenario.

Reply:
1. No, you also need `B @ C`, as you could use imports from `B` at negative levels if `ImplicitStagePersistence` is enabled.

```
module M where


...

b2 = $(lift b)
```

3. Since you need `M @ R` then you need `A @ R`, since inside `M` you might write definitions which use idenitifers from `A` at the top level. In fact, you do this and use `a` inside the definition of `Lift`. 

```
module Main where

import M

main = do
   let !x = lift (Some 42)
   print "done"
```

Clearly in this example you need `M @ R`, and therefore `A @ R`.

Reply:
> No, you also need B @ C, as you could use imports from B at negative levels if ImplicitStagePersistence is enabled.

Sorry, but I don't follow. Are you saying that, in the original program that I have given, if I need `C @ R`, I also need `B @ C`? The proposal currently says that, but I would like to know the reasoning for that, as I cannot see it.

I think we are perhaps talking past each other, so I tried coming up with more examples. Indeed, the following would be troubling if accepted:

```hs
module A where { a = 1 :: Int }

{-# LANGUAGE ExplicitLevelImports #-}
{-# LANGUAGE ImplicitStagePersistence #-}
module B where
import A

foo = a

bar = [| foo |]

{-# LANGUAGE ExplicitLevelImports #-}
module C where
import splice B
c :: Int
c = $(bar) 
```

because `foo` is implicitly persisted to quote level in `B` through `bar`. Suddenly, dependency analysis as specified would say `A` is needed only at `C`, whereas in truth it is needed at `R`. Furthermore, since `c = foo` after splicing, we also need `import B` (normally).

So indeed ImplicitStagePersistence as specified can break the semantics of ExplicitLevelImports as specified. Could you add such an example to the proposal?

Reply:
Consider this example.

```
module B where

b = ()

{-# LANGUAGE ImplicitStagePersistence #-}
module C where
import B

-- b is used at -1
c = $(lift b)
```

Here `C @ R` clearly requires `B @ C` but this is not specified by the level imports. I feel like I have just repeated my explanation, we could have a call if you want me to explain more so I can understand where the confusion is.